imgui.cpp

```diff
@@ -867,6 +867,7 @@ CODE
 #else
 #include <stdint.h>     // intptr_t
 #endif
+#include <atomic>
 
```

Comment:
However simple that's a problem since we never include C++ library headers. 
We just... can't trust the amount of things they may be dragging.
I am not sure what the proper solution for this is tbh.

Reply:
@ocornut 

I was also feeling bad when I wrote this line of code. I think the best solution is to _reimplement simple atomic using compiler intrinsic_. I will try to implementation this solution.

Let me explain how I came to that conclusion:

<details><summary>(click to expand)</summary>
<p>

#### Can't `ImGui::MemAlloc` and `ImGui::MemFree` remain context dependent?

From a software architecture point of view having `ImGui::MemAlloc` and `ImGui::MemFree` be context dependent is the best. But today it does not look practical in the case we want to use multiple contexts at the same time:
- FreeType integration and imstub_truetype.h depends on allocation IM_ALLOC and would need to become context aware. It is not straightforward, maybe even impossible for transparent FreeType integration.
- The vulkan backend depends on IM_ALLOC in code which should be unaware of ImGuiContext*, a global allocator should be used instead, but which one?
- class like `ImVector` would need to be given an explicit context at construction time

So I think `ImGui::MemAlloc` and `ImGui::MemFree` should be context independent.

#### Do we really need to track allocation?

If we stop tracking allocation it is a feature regression. Of course if we choose that solution it is easy. But I consider feature regression is not acceptable.

#### Tracking allocation + Global Allocator + Multiple Context = need for synchronization

##### Solution 1: User is forced to provide synchronization

We forward the problem to the user. It does not sound fair, and it break the promise Dear ImGui is easy to integrate.

##### Solution 2: Using `#include <atomic>`

This is the only solution if we want a true cross-platform solution, because the is the way atomic are exposed in C++. But if we do not want to include C++ header, we need some compromise.

##### Solution 3: Using platform primitive: windows.h, pthread, ...

If we do not want to include `<atomic>`, I do not think we can accept to include platform specific header. Also we would need to provide one implementation per platform, and would not be able to provide implementation for private platform like game console

##### Solution 4: Using compiler intrinsic (my preference)

It sounds to be the best compromise. Compiler intrinsic headers should not be as heavy as `<atomic>`, and by providing an implementation for Clang, GCC and MSVC we should cover the compiler of most of Dear ImGui users.

</p>
</details>